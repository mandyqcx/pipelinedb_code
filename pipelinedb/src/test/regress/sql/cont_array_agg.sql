-- array_sort function
CREATE FUNCTION
  ca_array_sort(
      array_vals_to_sort anyarray
  )
  RETURNS TABLE (
    sorted_array anyarray
  )
  AS $BODY$
    BEGIN
      RETURN QUERY SELECT
        ARRAY_AGG(val) AS sorted_array
      FROM
        (
          SELECT
            UNNEST(array_vals_to_sort) AS val
          ORDER BY
            val
        ) AS sorted_vals
      ;
    END;
  $BODY$
LANGUAGE plpgsql;

-- array_agg
CREATE FOREIGN TABLE cqobjectagg_stream (k text, v int) SERVER pipelinedb;

CREATE VIEW test_array_agg AS SELECT k::text, array_agg(v::integer) FROM cqobjectagg_stream GROUP BY k;
\d+ test_array_agg_mrel

INSERT INTO cqobjectagg_stream (k, v) VALUES ('x', 0), ('x', 1), ('x', 2), ('x', 3);
INSERT INTO cqobjectagg_stream (k, v) VALUES ('y', 0), ('y', 1);

SELECT k, ca_array_sort(array_agg) FROM test_array_agg ORDER BY k;

INSERT INTO cqobjectagg_stream (k, v) VALUES ('x', 4), ('y', 2), ('z', 10), ('z', 20);

SELECT k, ca_array_sort(array_agg) FROM test_array_agg ORDER BY k;

DROP VIEW test_array_agg;

CREATE VIEW test_array_agg AS SELECT array_agg(k::text) FROM cqobjectagg_stream;

INSERT INTO cqobjectagg_stream (k) VALUES ('hello'), ('world');
SELECT pg_sleep(0.1);
INSERT INTO cqobjectagg_stream (k) VALUES ('lol'), ('cat');

SELECT ca_array_sort(array_agg) FROM test_array_agg;
DROP FOREIGN TABLE cqobjectagg_stream CASCADE;

-- array_agg_array
CREATE FOREIGN TABLE cqobjectagg_stream (x int, y int) SERVER pipelinedb;

CREATE VIEW test_array_agg_array AS SELECT array_agg(ARRAY[x::int, y::int]) FROM cqobjectagg_stream;

INSERT INTO cqobjectagg_stream (x, y) VALUES (1, 11);
INSERT INTO cqobjectagg_stream (x, y) VALUES (2, 12);
SELECT pg_sleep(0.1);
INSERT INTO cqobjectagg_stream (x, y) VALUES (3, 13);

SELECT * FROM test_array_agg_array;

DROP FOREIGN TABLE cqobjectagg_stream CASCADE;