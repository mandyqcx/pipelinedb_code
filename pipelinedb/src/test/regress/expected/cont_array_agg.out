-- array_sort function
CREATE FUNCTION
  ca_array_sort(
      array_vals_to_sort anyarray
  )
  RETURNS TABLE (
    sorted_array anyarray
  )
  AS $BODY$
    BEGIN
      RETURN QUERY SELECT
        ARRAY_AGG(val) AS sorted_array
      FROM
        (
          SELECT
            UNNEST(array_vals_to_sort) AS val
          ORDER BY
            val
        ) AS sorted_vals
      ;
    END;
  $BODY$
LANGUAGE plpgsql;
-- array_agg
CREATE FOREIGN TABLE cqobjectagg_stream (k text, v int) SERVER pipelinedb;
CREATE VIEW test_array_agg AS SELECT k::text, array_agg(v::integer) FROM cqobjectagg_stream GROUP BY k;
\d+ test_array_agg_mrel
                             Table "public.test_array_agg_mrel"
  Column   |  Type  | Collation | Nullable | Default | Storage  | Stats target | Description 
-----------+--------+-----------+----------+---------+----------+--------------+-------------
 k         | text   |           |          |         | extended |              | 
 array_agg | bytea  |           |          |         | extended |              | 
 $pk       | bigint |           | not null |         | plain    |              | 
Indexes:
    "test_array_agg_mrel_pkey" PRIMARY KEY, btree ("$pk")
    "test_array_agg_mrel_expr_idx" btree (pipelinedb.hash_group(k))
Options: fillfactor=50

INSERT INTO cqobjectagg_stream (k, v) VALUES ('x', 0), ('x', 1), ('x', 2), ('x', 3);
INSERT INTO cqobjectagg_stream (k, v) VALUES ('y', 0), ('y', 1);
SELECT k, ca_array_sort(array_agg) FROM test_array_agg ORDER BY k;
 k | ca_array_sort 
---+---------------
 x | {0,1,2,3}
 y | {0,1}
(2 rows)

INSERT INTO cqobjectagg_stream (k, v) VALUES ('x', 4), ('y', 2), ('z', 10), ('z', 20);
SELECT k, ca_array_sort(array_agg) FROM test_array_agg ORDER BY k;
 k | ca_array_sort 
---+---------------
 x | {0,1,2,3,4}
 y | {0,1,2}
 z | {10,20}
(3 rows)

DROP VIEW test_array_agg;
CREATE VIEW test_array_agg AS SELECT array_agg(k::text) FROM cqobjectagg_stream;
INSERT INTO cqobjectagg_stream (k) VALUES ('hello'), ('world');
SELECT pg_sleep(0.1);
 pg_sleep 
----------
 
(1 row)

INSERT INTO cqobjectagg_stream (k) VALUES ('lol'), ('cat');
SELECT ca_array_sort(array_agg) FROM test_array_agg;
     ca_array_sort     
-----------------------
 {cat,hello,lol,world}
(1 row)

DROP FOREIGN TABLE cqobjectagg_stream CASCADE;
NOTICE:  drop cascades to view test_array_agg
-- array_agg_array
CREATE FOREIGN TABLE cqobjectagg_stream (x int, y int) SERVER pipelinedb;
CREATE VIEW test_array_agg_array AS SELECT array_agg(ARRAY[x::int, y::int]) FROM cqobjectagg_stream;
INSERT INTO cqobjectagg_stream (x, y) VALUES (1, 11);
INSERT INTO cqobjectagg_stream (x, y) VALUES (2, 12);
SELECT pg_sleep(0.1);
 pg_sleep 
----------
 
(1 row)

INSERT INTO cqobjectagg_stream (x, y) VALUES (3, 13);
SELECT * FROM test_array_agg_array;
       array_agg        
------------------------
 {{1,11},{2,12},{3,13}}
(1 row)

DROP FOREIGN TABLE cqobjectagg_stream CASCADE;
NOTICE:  drop cascades to view test_array_agg_array
